File: ProjectFinal/BE/ParkingManagement/DomainLayer/Interfaces/IGenericRepository.cs
using DataAccessLayer.Common;
using System.Linq.Expressions;

namespace DomainLayer.Interfaces
{
    public interface IGenericRepository<TEntity> where TEntity : class
    {
        Task<TEntity?> GetByIdAsync(int id);
        Task<IReadOnlyList<TEntity>> GetAllAsync();

        Task<IReadOnlyList<TEntity>> FindAsync(
            Expression<Func<TEntity, bool>> predicate);

        Task AddAsync(TEntity entity);
        void Update(TEntity entity);
        void Remove(TEntity entity);

        Task<bool> ExistsAsync(Expression<Func<TEntity, bool>> predicate);
        Task<PaginatedResult<TEntity>> GetPagedAsync(
                       int pageIndex,
                       int pageSize,
                       Expression<Func<TEntity, bool>>? filter = null,
                       Func<IQueryable<TEntity>, IOrderedQueryable<TEntity>>? orderBy = null,
                       string includeProperties = "");

    }
}

File: VBNet/MyLibrary.DAL/Repositories/Interfaces/IGenericRepository.vb
Imports System.Linq.Expressions

Public Interface IGenericRepository(Of T As Class)

    Function GetById(id As Integer) As T
    Function GetAll() As IQueryable(Of T)
    Function Find(predicate As Expression(Of Func(Of T, Boolean))) As IEnumerable(Of T)

    Sub Add(entity As T)
    Sub Update(entity As T)
    Sub SoftDelete(entity As T)

End Interface
